### PR TITLE
Visual Studio for Mac version protection

### PR DESCRIPTION
**Pull request by sfoslund**
Adding protection to ensure the tool does not uninstall bundles required by Visual Studio for Mac. Prevents (without the --force option) the user from uninstalling the highest installed major.minor runtime versions and the highest installed sdk version.

### CODE REVIEW

File: src/dotnet-core-uninstall/Shared/Configs/CommandLineConfigs.cs [hunk 1]
```diff
@@ -227,7 +228,9 @@ static CommandLineConfigs()
             DryRunCommand.Handler = CommandHandler.Create(ExceptionHandler.HandleException(() => DryRunCommandExec.Execute()));
             RemoveCommand.Handler = CommandHandler.Create(ExceptionHandler.HandleException(() => UninstallCommandExec.Execute()));
 
-            CommandLineParseResult = UninstallRootCommand.Parse(Environment.GetCommandLineArgs());
+            var parser = new CommandLineBuilder(UninstallRootCommand)
```
Comment (wli3) [inline src/dotnet-core-uninstall/Shared/Configs/CommandLineConfigs.cs]: Why this change? `CommandLineBuilder(UninstallRootCommand)` should be the default.

File: test/dotnet-core-uninstall.Tests/Shared/VSVersioning/VSVersionTests.cs [hunk 2]
```diff
@@ -64,97 +77,167 @@ internal void TestGetUninstallable(string[] versions, bool[] allowed)
...
-        [InlineData(new string[] { "5.0.0", "5.0.1", "10.100.100" }, new bool[] { true, true, true })]
-        internal void TestGetUninstallableNonSdkVersionsMac(string[] versions, bool[] allowed)
-        {
-            TestGetUninstallableNonSdkVersions(versions, allowed);
-        }
-
-        internal void TestGetUninstallableNonSdkVersions(string[] versions, bool[] allowed)
+        [InlineData(new string[] { "1.0.0" }, new bool[] { false }, new string[] { "1.0.0" }, new bool[] { false })]
+        [InlineData(new string[] { "1.0.0", "1.0.1" }, new bool[] { true, false }, new string[] { "1.0.0", "1.0.1" }, new bool[] { true, false })]
+        [InlineData(new string[] { "2.1.0", "1.0.1" }, new bool[] { false, true }, new string[] { "2.0.0", "1.1.0" }, new bool[] { false, false })]
```
Comment (wli3) [inline test/dotnet-core-uninstall.Tests/Shared/VSVersioning/VSVersionTests.cs]: new string[] { "2.1.0", "1.0.1" }, new bool[] { false, true }

why 2.1.0 is allowed but 1.01 is not? Did we only want 3.0.0 for VS for mac?

File: src/dotnet-core-uninstall/Shared/Configs/CommandLineConfigs.cs (hunk 1 above)
Comment (sfoslund) [reply src/dotnet-core-uninstall/Shared/Configs/CommandLineConfigs.cs]: I'm working on customizing the help builder right now which will require a parser build from the command line builder. I didn't mean to include this change in this commit, but it will be part of the next PR.

File: test/dotnet-core-uninstall.Tests/Shared/VSVersioning/VSVersionTests.cs (hunk 2 above)
Comment (sfoslund) [reply test/dotnet-core-uninstall.Tests/Shared/VSVersioning/VSVersionTests.cs]: VS for mac only needs the highest SDK, so we only ever have to keep one. That test is saying that we should be able to uninstall 1.0.1 (it is allowed to be uninstalled) but 2.1.0 is required by VS for mac since it's the highest SDK "installed"

File: src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs [hunk 3]
```diff
@@ -41,13 +41,43 @@ private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) A
...
+        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyMacVersionDivisions(IEnumerable<Bundle> bundleList)
         {
-            if (!RuntimeInfo.RunningOnWindows)
+            var bundlesAboveLimit = bundleList.Where(bundle => bundle.Version.SemVer >= UpperLimit);
+            bundleList = bundleList.Except(bundlesAboveLimit);
+
+            var dividedBundles = bundleList
+                .Where(bundle => bundle.Version is RuntimeVersion)
+                .GroupBy(bundle => bundle.Version.MajorMinor)
+                .Select(pair => (pair as IEnumerable<Bundle>, string.Format(LocalizableStrings.RequirementExplainationString, " for Mac or SDKs")))
```
[the remark this answers is not in the document]
Comment (sfoslund) [reply src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs]: Not totally sure what you mean... Are you suggesting using a Mac specific version of the RequirementExplainationString? That seems like a good idea.

[the remark this answers is not in the document]
Comment (wli3) [reply src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs]: "Maybe needed for Visual Studio" + " for Mac or SDKs" when translated separately to different languages and then join together probably won't all make sense.

Might just have "Maybe needed for Visual Studio for Mac or SDKs" as one string
